feat: Implement Scrapy HTTP cache backend

honzajavorek opened the pull request:
I successfully use this code in my project. In 381c0445067f97940e5c0b01fa83c3ab4216f946 I explicitly specify licensing. I didn't add docs or anything else (yet), just made sure the code passes linters and a type check.

Relates: https://github.com/apify/actor-templates/issues/303

vdusek commented on src/apify/scrapy/cache.py:
```diff
@@ -0,0 +1,176 @@
+from __future__ import annotations
```
Taking account the current project structure and the fact that this is an [Extension](https://docs.scrapy.org/en/latest/topics/extensions.html) (correct me if I am wrong), add it please to the `extensions/` subpackage with `__init__.py` exposing it.

vdusek replied:
`extensions/_httpcache.py` and `extensions/__init__.py` exposing `ApifyCacheStorage` please

honzajavorek replied:
So the recommended way to import the class is `apify.scrapy.extensions.httpcache.ApifyStorage` or `apify.scrapy.extensions.ApifyStorage`? So that I know how to adjust docs and the module path in the apply settings function.

vdusek commented on src/apify/scrapy/cache.py:
```diff
@@ -0,0 +1,176 @@
+    """
+
+    def __init__(self, settings: BaseSettings) -> None:
+        self.expiration_max_items = 100
+        self.expiration_secs: int = settings.getint('HTTPCACHE_EXPIRATION_SECS')
```
If these have to be public, maybe we can use just read-only properties?

honzajavorek replied:
Addressed in 224fca5

vdusek commented on src/apify/scrapy/cache.py:
```diff
@@ -0,0 +1,176 @@
+logger = getLogger(__name__)
+
+
+class ApifyCacheStorage:
```
Isn't this specifically HTTP cache storage? (rather then general cache)

honzajavorek replied:
The built-in Scrapy cache storages are called `DbmCacheStorage` or `FilesystemCacheStorage`. They live in a module `scrapy.extensions.httpcache`. Maybe I could rename the module to `httpcache`.

honzajavorek replied:
Addressed in 3a2cc13

vdusek replied:
Makes sense, thanks.

vdusek commented on src/apify/scrapy/cache.py:
```diff
@@ -0,0 +1,176 @@
+
+class ApifyCacheStorage:
+    """A Scrapy cache storage that uses the Apify `KeyValueStore` to store responses.
+
```
Could you add that it is being used by the Scrapy's `HttpCacheMiddleware`?

honzajavorek replied:
Addressed in 155caf2

vdusek commented on src/apify/scrapy/cache.py:
```diff
@@ -0,0 +1,176 @@
+        self._fingerprinter = spider.crawler.request_fingerprinter
+        kv_name = f'httpcache-{spider.name}'
+
+        async def open_kv() -> KeyValueStore:
```
```suggestion
        async def open_kvs() -> KeyValueStore:
```

we use "KVS" shortcut for key-value store.

honzajavorek replied:
Addressed in 3909758

vdusek reviewed (changes requested):
To make this work by default, also adjust the `apify.scrapy.utils.apply_apify_settings`. If this allows us to handle platform migration, we should turn caching by default. Also please add a reasonable comments there.

honzajavorek commented:
Thanks for the review! I considered this to be a kick off which will definitely get a review with changes requested. If https://github.com/apify/apify-sdk-python/issues/404 lets me, I'll work on addressing the comments in the upcoming days.

honzajavorek commented:
I bet I've seen comments about adding this to Apify settings by default, but I can't see them now. I think it's a good idea to set the cache storage by default, but I wouldn't turn on caching by default, because it's application-specific how caching should work exactly (e.g. expiration time). We should just document somewhere that this is a preferred solution to handling the forced restarts, if they happen.

vdusek commented:
> I bet I've seen comments about adding this to Apify settings by default, but I can't see them now. I think it's a good idea to set the cache storage by default, but I wouldn't turn on caching by default, because it's application-specific how caching should work exactly (e.g. expiration time). We should just document somewhere that this is a preferred solution to handling the forced restarts, if they happen.

Agreed, thanks :slightly_smiling_face:.

honzajavorek commented:
Just for the record, the comment I've seen didn't get lost, it's the [final comment of the review](https://github.com/apify/apify-sdk-python/pull/403#pullrequestreview-2623112514) 😄

honzajavorek commented on src/apify/scrapy/cache.py:
```diff
@@ -0,0 +1,176 @@
+        logger.debug('Using Apify key value cache storage', extra={'spider': spider})
+        self.spider = spider
+        self._fingerprinter = spider.crawler.request_fingerprinter
+        kv_name = f'httpcache-{spider.name}'
```
Addressed in 3909758

honzajavorek commented on src/apify/scrapy/cache.py:
```diff
@@ -0,0 +1,176 @@
+        logger.debug("Starting background thread for cache storage's event loop")
+        self._async_thread = AsyncThread()
+        logger.debug(f"Opening cache storage's {kv_name!r} key value store")
+        self._kv = self._async_thread.run_coro(open_kv())
```
Addressed in 3909758

honzajavorek commented:
I think I'm done for now!

vdusek commented on docs/02_guides/05_scrapy.mdx:
```diff
@@ -94,6 +95,12 @@ The following example demonstrates a Scrapy Actor that scrapes page titles and e
     </TabItem>
 </Tabs>
 
+## Dealing with ‘imminent migration to another host’
```
Probably `'` or `"` 
```suggestion
## Dealing with 'imminent migration to another host'
```

vdusek reviewed (changes requested):
Error from the [integration test](https://github.com/apify/apify-sdk-python/blob/master/tests/integration/test_actor_scrapy.py):

```
2025-03-14T11:20:46.570Z ACTOR: Pulling Docker image of build dtQ5GdsNexthjANND from repository.
2025-03-14T11:20:50.429Z ACTOR: Creating Docker container.
2025-03-14T11:20:50.518Z ACTOR: Starting Docker container.
2025-03-14T11:20:52.574Z [apify] INFO  Initializing Actor...
2025-03-14T11:20:52.574Z [apify] INFO  Initializing Actor... ({"message": "Initializing Actor..."})
2025-03-14T11:20:52.576Z [apify] INFO  System info ({"apify_sdk_version": "2.4.0", "apify_client_version": "1.9.2", "crawlee_version": "0.6.5", "python_version": "3.13.2", "os": "linux"})
2025-03-14T11:20:52.577Z [apify] INFO  System info ({"apify_sdk_version": "2.4.0", "apify_client_version": "1.9.2", "crawlee_version": "0.6.5", "python_version": "3.13.2", "os": "linux", "message": "System info"})
2025-03-14T11:20:52.692Z [scrapy.addons] INFO  Enabled addons:
2025-03-14T11:20:52.695Z [] ({"crawler": "<scrapy.crawler.Crawler object at 0x7a71975fdbe0>"})
2025-03-14T11:20:52.742Z [scrapy.middleware] INFO  Enabled extensions:
2025-03-14T11:20:52.743Z ['scrapy.extensions.corestats.CoreStats',
2025-03-14T11:20:52.743Z  'scrapy.extensions.memusage.MemoryUsage',
2025-03-14T11:20:52.744Z  'scrapy.extensions.closespider.CloseSpider',
2025-03-14T11:20:52.745Z  'scrapy.extensions.logstats.LogStats'] ({"crawler": "<scrapy.crawler.Crawler object at 0x7a71975fdbe0>"})
2025-03-14T11:20:52.745Z [scrapy.crawler] INFO  Overridden settings:
2025-03-14T11:20:52.746Z {'BOT_NAME': 'titlebot',
2025-03-14T11:20:52.747Z  'CLOSESPIDER_PAGECOUNT': 10,
2025-03-14T11:20:52.749Z  'DEPTH_LIMIT': 1,
2025-03-14T11:20:52.750Z  'HTTPCACHE_ENABLED': True,
2025-03-14T11:20:52.751Z  'HTTPCACHE_EXPIRATION_SECS': 7200,
2025-03-14T11:20:52.752Z  'HTTPCACHE_STORAGE': 'apify.scrapy.extensions.ApifyCacheStorage',
2025-03-14T11:20:52.753Z  'LOG_LEVEL': 'INFO',
2025-03-14T11:20:52.754Z  'NEWSPIDER_MODULE': 'src.spiders',
2025-03-14T11:20:52.755Z  'ROBOTSTXT_OBEY': True,
2025-03-14T11:20:52.756Z  'SCHEDULER': 'apify.scrapy.scheduler.ApifyScheduler',
2025-03-14T11:20:52.757Z  'SPIDER_MODULES': ['src.spiders'],
2025-03-14T11:20:52.758Z  'TELNETCONSOLE_ENABLED': False,
2025-03-14T11:20:52.761Z  'TWISTED_REACTOR': 'twisted.internet.asyncioreactor.AsyncioSelectorReactor'}
2025-03-14T11:20:52.834Z [scrapy.middleware] INFO  Enabled downloader middlewares:
2025-03-14T11:20:52.835Z ['scrapy.downloadermiddlewares.offsite.OffsiteMiddleware',
2025-03-14T11:20:52.836Z  'scrapy.downloadermiddlewares.robotstxt.RobotsTxtMiddleware',
2025-03-14T11:20:52.837Z  'scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
2025-03-14T11:20:52.838Z  'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
2025-03-14T11:20:52.839Z  'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
2025-03-14T11:20:52.840Z  'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
2025-03-14T11:20:52.840Z  'scrapy.downloadermiddlewares.retry.RetryMiddleware',
2025-03-14T11:20:52.841Z  'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
2025-03-14T11:20:52.842Z  'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
2025-03-14T11:20:52.842Z  'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
2025-03-14T11:20:52.843Z  'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
2025-03-14T11:20:52.844Z  'apify.scrapy.middlewares.ApifyHttpProxyMiddleware',
2025-03-14T11:20:52.844Z  'scrapy.downloadermiddlewares.stats.DownloaderStats',
2025-03-14T11:20:52.845Z  'scrapy.downloadermiddlewares.httpcache.HttpCacheMiddleware'] ({"crawler": "<scrapy.crawler.Crawler object at 0x7a71975fdbe0>"})
2025-03-14T11:20:52.846Z [scrapy.middleware] INFO  Enabled spider middlewares:
2025-03-14T11:20:52.847Z ['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
2025-03-14T11:20:52.848Z  'scrapy.spidermiddlewares.referer.RefererMiddleware',
2025-03-14T11:20:52.848Z  'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
2025-03-14T11:20:52.849Z  'scrapy.spidermiddlewares.depth.DepthMiddleware'] ({"crawler": "<scrapy.crawler.Crawler object at 0x7a71975fdbe0>"})
2025-03-14T11:20:52.850Z [scrapy.middleware] INFO  Enabled item pipelines:
2025-03-14T11:20:52.851Z ['apify.scrapy.pipelines.ActorDatasetPushPipeline'] ({"crawler": "<scrapy.crawler.Crawler object at 0x7a71975fdbe0>"})
2025-03-14T11:20:52.852Z [scrapy.core.engine] INFO  Spider opened ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:52.959Z [scrapy.extensions.logstats] INFO  Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min) ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.023Z [apify.scrapy._async_thread] ERROR Coroutine execution raised an exception.
2025-03-14T11:20:53.024Z       Traceback (most recent call last):
2025-03-14T11:20:53.025Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/_async_thread.py", line 62, in run_coro
2025-03-14T11:20:53.027Z           return future.result(timeout=timeout.total_seconds())
2025-03-14T11:20:53.027Z                  ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.028Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 456, in result
2025-03-14T11:20:53.029Z           return self.__get_result()
2025-03-14T11:20:53.029Z                  ~~~~~~~~~~~~~~~~~^^
2025-03-14T11:20:53.030Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 401, in __get_result
2025-03-14T11:20:53.031Z           raise self._exception
2025-03-14T11:20:53.031Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 56, in open_kvs
2025-03-14T11:20:53.032Z           return await KeyValueStore.open(name=kvs_name, storage_client=storage_client)
2025-03-14T11:20:53.033Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.034Z         File "/usr/local/lib/python3.13/site-packages/crawlee/storages/_key_value_store.py", line 129, in open
2025-03-14T11:20:53.035Z           return await open_storage(
2025-03-14T11:20:53.035Z                  ^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.036Z           ...<5 lines>...
2025-03-14T11:20:53.036Z           )
2025-03-14T11:20:53.037Z           ^
2025-03-14T11:20:53.038Z         File "/usr/local/lib/python3.13/site-packages/crawlee/storages/_creation_management.py", line 170, in open_storage
2025-03-14T11:20:53.038Z           storage_object = await resource_collection_client.get_or_create(name=name)
2025-03-14T11:20:53.039Z                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.040Z         File "/usr/local/lib/python3.13/site-packages/apify/apify_storage_client/_key_value_store_collection_client.py", line 29, in get_or_create
2025-03-14T11:20:53.040Z           await self._client.get_or_create(
2025-03-14T11:20:53.043Z           ...<2 lines>...
2025-03-14T11:20:53.043Z           )
2025-03-14T11:20:53.044Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_logging.py", line 61, in async_wrapper
2025-03-14T11:20:53.045Z           return await fun(resource_client, *args, **kwargs)
2025-03-14T11:20:53.046Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.046Z         File "/usr/local/lib/python3.13/site-packages/apify_client/clients/resource_clients/key_value_store_collection.py", line 112, in get_or_create
2025-03-14T11:20:53.047Z           return await self._get_or_create(name=name, resource=filter_out_none_values_recursively({'schema': schema}))
2025-03-14T11:20:53.048Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.050Z         File "/usr/local/lib/python3.13/site-packages/apify_client/clients/base/resource_collection_client.py", line 74, in _get_or_create
2025-03-14T11:20:53.051Z           response = await self.http_client.call(
2025-03-14T11:20:53.052Z                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.052Z           ...<4 lines>...
2025-03-14T11:20:53.053Z           )
2025-03-14T11:20:53.054Z           ^
2025-03-14T11:20:53.054Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_http_client.py", line 286, in call
2025-03-14T11:20:53.055Z           return await retry_with_exp_backoff_async(
2025-03-14T11:20:53.056Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.056Z           ...<5 lines>...
2025-03-14T11:20:53.057Z           )
2025-03-14T11:20:53.058Z           ^
2025-03-14T11:20:53.058Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_utils.py", line 99, in retry_with_exp_backoff_async
2025-03-14T11:20:53.059Z           return await async_func(stop_retrying, attempt)
2025-03-14T11:20:53.059Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.060Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_http_client.py", line 284, in _make_request
2025-03-14T11:20:53.061Z           raise ApifyApiError(response, attempt)
2025-03-14T11:20:53.061Z       apify_client._errors.ApifyApiError: Invalid value provided in store: name can only contain letters 'a' through 'z', the digits '0' through '9', and the hyphen ('-') but only in the middle of the string (e.g. 'my-value-1')
2025-03-14T11:20:53.062Z [apify.scrapy._async_thread] ERROR Coroutine execution raised an exception. ({"message": "Coroutine execution raised an exception."})
2025-03-14T11:20:53.063Z       Traceback (most recent call last):
2025-03-14T11:20:53.064Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/_async_thread.py", line 62, in run_coro
2025-03-14T11:20:53.064Z           return future.result(timeout=timeout.total_seconds())
2025-03-14T11:20:53.065Z                  ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.066Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 456, in result
2025-03-14T11:20:53.067Z           return self.__get_result()
2025-03-14T11:20:53.068Z                  ~~~~~~~~~~~~~~~~~^^
2025-03-14T11:20:53.069Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 401, in __get_result
2025-03-14T11:20:53.070Z           raise self._exception
2025-03-14T11:20:53.070Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 56, in open_kvs
2025-03-14T11:20:53.071Z           return await KeyValueStore.open(name=kvs_name, storage_client=storage_client)
2025-03-14T11:20:53.072Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.073Z         File "/usr/local/lib/python3.13/site-packages/crawlee/storages/_key_value_store.py", line 129, in open
2025-03-14T11:20:53.073Z           return await open_storage(
2025-03-14T11:20:53.074Z                  ^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.075Z           ...<5 lines>...
2025-03-14T11:20:53.076Z           )
2025-03-14T11:20:53.076Z           ^
2025-03-14T11:20:53.077Z         File "/usr/local/lib/python3.13/site-packages/crawlee/storages/_creation_management.py", line 170, in open_storage
2025-03-14T11:20:53.078Z           storage_object = await resource_collection_client.get_or_create(name=name)
2025-03-14T11:20:53.078Z                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.079Z         File "/usr/local/lib/python3.13/site-packages/apify/apify_storage_client/_key_value_store_collection_client.py", line 29, in get_or_create
2025-03-14T11:20:53.080Z           await self._client.get_or_create(
2025-03-14T11:20:53.081Z           ...<2 lines>...
2025-03-14T11:20:53.082Z           )
2025-03-14T11:20:53.083Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_logging.py", line 61, in async_wrapper
2025-03-14T11:20:53.083Z           return await fun(resource_client, *args, **kwargs)
2025-03-14T11:20:53.084Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.085Z         File "/usr/local/lib/python3.13/site-packages/apify_client/clients/resource_clients/key_value_store_collection.py", line 112, in get_or_create
2025-03-14T11:20:53.086Z           return await self._get_or_create(name=name, resource=filter_out_none_values_recursively({'schema': schema}))
2025-03-14T11:20:53.087Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.088Z         File "/usr/local/lib/python3.13/site-packages/apify_client/clients/base/resource_collection_client.py", line 74, in _get_or_create
2025-03-14T11:20:53.088Z           response = await self.http_client.call(
2025-03-14T11:20:53.089Z                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.090Z           ...<4 lines>...
2025-03-14T11:20:53.090Z           )
2025-03-14T11:20:53.091Z           ^
2025-03-14T11:20:53.092Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_http_client.py", line 286, in call
2025-03-14T11:20:53.092Z           return await retry_with_exp_backoff_async(
2025-03-14T11:20:53.093Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.093Z           ...<5 lines>...
2025-03-14T11:20:53.094Z           )
2025-03-14T11:20:53.095Z           ^
2025-03-14T11:20:53.096Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_utils.py", line 99, in retry_with_exp_backoff_async
2025-03-14T11:20:53.096Z           return await async_func(stop_retrying, attempt)
2025-03-14T11:20:53.097Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.097Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_http_client.py", line 284, in _make_request
2025-03-14T11:20:53.098Z           raise ApifyApiError(response, attempt)
2025-03-14T11:20:53.099Z       apify_client._errors.ApifyApiError: Invalid value provided in store: name can only contain letters 'a' through 'z', the digits '0' through '9', and the hyphen ('-') but only in the middle of the string (e.g. 'my-value-1')
2025-03-14T11:20:53.101Z [scrapy.utils.signal] ERROR Error caught on signal handler: <bound method HttpCacheMiddleware.spider_opened of <scrapy.downloadermiddlewares.httpcache.HttpCacheMiddleware object at 0x7a719718a510>> ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.102Z       Traceback (most recent call last):
2025-03-14T11:20:53.102Z         File "/usr/local/lib/python3.13/site-packages/scrapy/utils/defer.py", line 400, in maybeDeferred_coro
2025-03-14T11:20:53.103Z           result = f(*args, **kw)
2025-03-14T11:20:53.104Z         File "/usr/local/lib/python3.13/site-packages/pydispatch/robustapply.py", line 55, in robustApply
2025-03-14T11:20:53.104Z           return receiver(*arguments, **named)
2025-03-14T11:20:53.105Z         File "/usr/local/lib/python3.13/site-packages/scrapy/downloadermiddlewares/httpcache.py", line 65, in spider_opened
2025-03-14T11:20:53.106Z           self.storage.open_spider(spider)
2025-03-14T11:20:53.106Z           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
2025-03-14T11:20:53.107Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 62, in open_spider
2025-03-14T11:20:53.108Z           self._kvs = self._async_thread.run_coro(open_kvs())
2025-03-14T11:20:53.109Z                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
2025-03-14T11:20:53.110Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/_async_thread.py", line 62, in run_coro
2025-03-14T11:20:53.111Z           return future.result(timeout=timeout.total_seconds())
2025-03-14T11:20:53.112Z                  ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.113Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 456, in result
2025-03-14T11:20:53.113Z           return self.__get_result()
2025-03-14T11:20:53.114Z                  ~~~~~~~~~~~~~~~~~^^
2025-03-14T11:20:53.115Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 401, in __get_result
2025-03-14T11:20:53.116Z           raise self._exception
2025-03-14T11:20:53.117Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 56, in open_kvs
2025-03-14T11:20:53.118Z           return await KeyValueStore.open(name=kvs_name, storage_client=storage_client)
2025-03-14T11:20:53.119Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.120Z         File "/usr/local/lib/python3.13/site-packages/crawlee/storages/_key_value_store.py", line 129, in open
2025-03-14T11:20:53.120Z           return await open_storage(
2025-03-14T11:20:53.121Z                  ^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.122Z           ...<5 lines>...
2025-03-14T11:20:53.122Z           )
2025-03-14T11:20:53.123Z           ^
2025-03-14T11:20:53.124Z         File "/usr/local/lib/python3.13/site-packages/crawlee/storages/_creation_management.py", line 170, in open_storage
2025-03-14T11:20:53.124Z           storage_object = await resource_collection_client.get_or_create(name=name)
2025-03-14T11:20:53.125Z                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.126Z         File "/usr/local/lib/python3.13/site-packages/apify/apify_storage_client/_key_value_store_collection_client.py", line 29, in get_or_create
2025-03-14T11:20:53.126Z           await self._client.get_or_create(
2025-03-14T11:20:53.127Z           ...<2 lines>...
2025-03-14T11:20:53.128Z           )
2025-03-14T11:20:53.129Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_logging.py", line 61, in async_wrapper
2025-03-14T11:20:53.129Z           return await fun(resource_client, *args, **kwargs)
2025-03-14T11:20:53.130Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.133Z         File "/usr/local/lib/python3.13/site-packages/apify_client/clients/resource_clients/key_value_store_collection.py", line 112, in get_or_create
2025-03-14T11:20:53.133Z           return await self._get_or_create(name=name, resource=filter_out_none_values_recursively({'schema': schema}))
2025-03-14T11:20:53.134Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.135Z         File "/usr/local/lib/python3.13/site-packages/apify_client/clients/base/resource_collection_client.py", line 74, in _get_or_create
2025-03-14T11:20:53.136Z           response = await self.http_client.call(
2025-03-14T11:20:53.137Z                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.137Z           ...<4 lines>...
2025-03-14T11:20:53.138Z           )
2025-03-14T11:20:53.139Z           ^
2025-03-14T11:20:53.139Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_http_client.py", line 286, in call
2025-03-14T11:20:53.140Z           return await retry_with_exp_backoff_async(
2025-03-14T11:20:53.141Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.142Z           ...<5 lines>...
2025-03-14T11:20:53.142Z           )
2025-03-14T11:20:53.143Z           ^
2025-03-14T11:20:53.144Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_utils.py", line 99, in retry_with_exp_backoff_async
2025-03-14T11:20:53.145Z           return await async_func(stop_retrying, attempt)
2025-03-14T11:20:53.145Z                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.146Z         File "/usr/local/lib/python3.13/site-packages/apify_client/_http_client.py", line 284, in _make_request
2025-03-14T11:20:53.147Z           raise ApifyApiError(response, attempt)
2025-03-14T11:20:53.148Z       apify_client._errors.ApifyApiError: Invalid value provided in store: name can only contain letters 'a' through 'z', the digits '0' through '9', and the hyphen ('-') but only in the middle of the string (e.g. 'my-value-1')
2025-03-14T11:20:53.499Z [scrapy.downloadermiddlewares.robotstxt] ERROR Error downloading <GET https://crawlee.dev/robots.txt>: Key value store not initialized ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.501Z       Traceback (most recent call last):
2025-03-14T11:20:53.502Z         File "/usr/local/lib/python3.13/site-packages/twisted/internet/defer.py", line 2017, in _inlineCallbacks
2025-03-14T11:20:53.502Z           result = context.run(gen.send, result)
2025-03-14T11:20:53.503Z         File "/usr/local/lib/python3.13/site-packages/scrapy/core/downloader/middleware.py", line 57, in process_request
2025-03-14T11:20:53.504Z           method(request=request, spider=spider)
2025-03-14T11:20:53.504Z           ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.505Z         File "/usr/local/lib/python3.13/site-packages/scrapy/downloadermiddlewares/httpcache.py", line 82, in process_request
2025-03-14T11:20:53.506Z           cachedresponse: Response | None = self.storage.retrieve_response(
2025-03-14T11:20:53.506Z                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
2025-03-14T11:20:53.507Z               spider, request
2025-03-14T11:20:53.508Z               ^^^^^^^^^^^^^^^
2025-03-14T11:20:53.509Z           )
2025-03-14T11:20:53.509Z           ^
2025-03-14T11:20:53.510Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 112, in retrieve_response
2025-03-14T11:20:53.511Z           raise ValueError('Key value store not initialized')
2025-03-14T11:20:53.513Z       ValueError: Key value store not initialized
2025-03-14T11:20:53.637Z [scrapy.core.scraper] ERROR Error downloading <GET https://crawlee.dev> ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.638Z       Traceback (most recent call last):
2025-03-14T11:20:53.639Z         File "/usr/local/lib/python3.13/site-packages/twisted/internet/defer.py", line 2017, in _inlineCallbacks
2025-03-14T11:20:53.639Z           result = context.run(gen.send, result)
2025-03-14T11:20:53.640Z         File "/usr/local/lib/python3.13/site-packages/scrapy/core/downloader/middleware.py", line 57, in process_request
2025-03-14T11:20:53.641Z           method(request=request, spider=spider)
2025-03-14T11:20:53.641Z           ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.642Z         File "/usr/local/lib/python3.13/site-packages/scrapy/downloadermiddlewares/httpcache.py", line 82, in process_request
2025-03-14T11:20:53.643Z           cachedresponse: Response | None = self.storage.retrieve_response(
2025-03-14T11:20:53.644Z                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
2025-03-14T11:20:53.645Z               spider, request
2025-03-14T11:20:53.645Z               ^^^^^^^^^^^^^^^
2025-03-14T11:20:53.648Z           )
2025-03-14T11:20:53.648Z           ^
2025-03-14T11:20:53.649Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 112, in retrieve_response
2025-03-14T11:20:53.650Z           raise ValueError('Key value store not initialized')
2025-03-14T11:20:53.651Z       ValueError: Key value store not initialized
2025-03-14T11:20:53.865Z [scrapy.core.engine] INFO  Closing spider (finished) ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.866Z [apify.scrapy.extensions._httpcache] INFO  Cleaning up cache items (max 100)
2025-03-14T11:20:53.868Z [apify.scrapy.extensions._httpcache] INFO  Cleaning up cache items (max 100) ({"message": "Cleaning up cache items (max 100)"})
2025-03-14T11:20:53.869Z [apify.scrapy._async_thread] ERROR Coroutine execution raised an exception.
2025-03-14T11:20:53.870Z       Traceback (most recent call last):
2025-03-14T11:20:53.870Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/_async_thread.py", line 62, in run_coro
2025-03-14T11:20:53.873Z           return future.result(timeout=timeout.total_seconds())
2025-03-14T11:20:53.874Z                  ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.874Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 456, in result
2025-03-14T11:20:53.875Z           return self.__get_result()
2025-03-14T11:20:53.876Z                  ~~~~~~~~~~~~~~~~~^^
2025-03-14T11:20:53.876Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 401, in __get_result
2025-03-14T11:20:53.877Z           raise self._exception
2025-03-14T11:20:53.877Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 76, in expire_kvs
2025-03-14T11:20:53.878Z           raise ValueError('Key value store not initialized')
2025-03-14T11:20:53.879Z       ValueError: Key value store not initialized
2025-03-14T11:20:53.879Z [apify.scrapy._async_thread] ERROR Coroutine execution raised an exception. ({"message": "Coroutine execution raised an exception."})
2025-03-14T11:20:53.880Z       Traceback (most recent call last):
2025-03-14T11:20:53.881Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/_async_thread.py", line 62, in run_coro
2025-03-14T11:20:53.881Z           return future.result(timeout=timeout.total_seconds())
2025-03-14T11:20:53.882Z                  ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.883Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 456, in result
2025-03-14T11:20:53.883Z           return self.__get_result()
2025-03-14T11:20:53.884Z                  ~~~~~~~~~~~~~~~~~^^
2025-03-14T11:20:53.885Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 401, in __get_result
2025-03-14T11:20:53.885Z           raise self._exception
2025-03-14T11:20:53.886Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 76, in expire_kvs
2025-03-14T11:20:53.887Z           raise ValueError('Key value store not initialized')
2025-03-14T11:20:53.887Z       ValueError: Key value store not initialized
2025-03-14T11:20:53.888Z [scrapy.utils.signal] ERROR Error caught on signal handler: <bound method HttpCacheMiddleware.spider_closed of <scrapy.downloadermiddlewares.httpcache.HttpCacheMiddleware object at 0x7a719718a510>> ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.889Z       Traceback (most recent call last):
2025-03-14T11:20:53.889Z         File "/usr/local/lib/python3.13/site-packages/scrapy/utils/defer.py", line 400, in maybeDeferred_coro
2025-03-14T11:20:53.890Z           result = f(*args, **kw)
2025-03-14T11:20:53.891Z         File "/usr/local/lib/python3.13/site-packages/pydispatch/robustapply.py", line 55, in robustApply
2025-03-14T11:20:53.891Z           return receiver(*arguments, **named)
2025-03-14T11:20:53.892Z         File "/usr/local/lib/python3.13/site-packages/scrapy/downloadermiddlewares/httpcache.py", line 68, in spider_closed
2025-03-14T11:20:53.893Z           self.storage.close_spider(spider)
2025-03-14T11:20:53.894Z           ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
2025-03-14T11:20:53.894Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 95, in close_spider
2025-03-14T11:20:53.895Z           self._async_thread.run_coro(expire_kvs())
2025-03-14T11:20:53.896Z           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
2025-03-14T11:20:53.896Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/_async_thread.py", line 62, in run_coro
2025-03-14T11:20:53.897Z           return future.result(timeout=timeout.total_seconds())
2025-03-14T11:20:53.898Z                  ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-03-14T11:20:53.898Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 456, in result
2025-03-14T11:20:53.899Z           return self.__get_result()
2025-03-14T11:20:53.899Z                  ~~~~~~~~~~~~~~~~~^^
2025-03-14T11:20:53.900Z         File "/usr/local/lib/python3.13/concurrent/futures/_base.py", line 401, in __get_result
2025-03-14T11:20:53.901Z           raise self._exception
2025-03-14T11:20:53.901Z         File "/usr/local/lib/python3.13/site-packages/apify/scrapy/extensions/_httpcache.py", line 76, in expire_kvs
2025-03-14T11:20:53.902Z           raise ValueError('Key value store not initialized')
2025-03-14T11:20:53.903Z       ValueError: Key value store not initialized
2025-03-14T11:20:53.904Z [scrapy.statscollectors] INFO  Dumping Scrapy stats:
2025-03-14T11:20:53.905Z {'downloader/exception_count': 2,
2025-03-14T11:20:53.905Z  'downloader/exception_type_count/builtins.ValueError': 2,
2025-03-14T11:20:53.906Z  'downloader/request_bytes': 680,
2025-03-14T11:20:53.906Z  'downloader/request_count': 2,
2025-03-14T11:20:53.907Z  'downloader/request_method_count/GET': 2,
2025-03-14T11:20:53.908Z  'elapsed_time_seconds': 0.906427,
2025-03-14T11:20:53.908Z  'finish_reason': 'finished',
2025-03-14T11:20:53.909Z  'finish_time': datetime.datetime(2025, 3, 14, 11, 20, 53, 865764, tzinfo=datetime.timezone.utc),
2025-03-14T11:20:53.910Z  'items_per_minute': None,
2025-03-14T11:20:53.912Z  'log_count/ERROR': 6,
2025-03-14T11:20:53.913Z  'log_count/INFO': 9,
2025-03-14T11:20:53.913Z  'log_count/WARNING': 1,
2025-03-14T11:20:53.914Z  'memusage/max': 130392064,
2025-03-14T11:20:53.915Z  'memusage/startup': 130392064,
2025-03-14T11:20:53.915Z  'responses_per_minute': None,
2025-03-14T11:20:53.916Z  "robotstxt/exception_count/<class 'ValueError'>": 1,
2025-03-14T11:20:53.917Z  'robotstxt/request_count': 1,
2025-03-14T11:20:53.917Z  'start_time': datetime.datetime(2025, 3, 14, 11, 20, 52, 959337, tzinfo=datetime.timezone.utc)} ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.918Z [scrapy.core.engine] INFO  Spider closed (finished) ({"spider": "<TitleSpider 'title_spider' at 0x7a71975fd7f0>"})
2025-03-14T11:20:53.919Z [apify] INFO  Exiting Actor ({"exit_code": 0})
2025-03-14T11:20:53.920Z [apify] INFO  Exiting Actor ({"exit_code": 0, "message": "Exiting Actor"})
```

vdusek commented on src/apify/scrapy/extensions/_httpcache.py:
```diff
@@ -0,0 +1,179 @@
+        logger.debug('Using Apify key value cache storage', extra={'spider': spider})
+        self._spider = spider
+        self._fingerprinter = spider.crawler.request_fingerprinter
+        kvs_name = f'httpcache-{spider.name}'
```
It looks like the KVS name cannot contain underscores.

honzajavorek replied:
Ha! My spiders apparently always had names with dashes. Can I bring in a dependency on `python-slugify`? Or maybe that's an overkill? I'll check what are allowed characters for spiders in Scrapy. Maybe some sort of replace would suffice.

vdusek replied:
Please do not use any other 3rd party library, that is not currently installed.

honzajavorek replied:
Addressed in [4ff56d7](https://github.com/apify/apify-sdk-python/pull/403/commits/4ff56d7ebc57ec9cfb22cf201276a8e63b8e6d74). I opted for simplicity - the KVS name is determined from the spider name in a naive way, but I think it should suffice to most people. One thing I want to keep is to be able to inspect the KVS in the Console and immediately know to which scraper it belongs:

<img width="1101" alt="Screenshot 2025-03-18 at 9 27 31" src="https://github.com/user-attachments/assets/0a8d47d5-4e57-4909-a07f-62070fe8d68c" />

I don't anticipate that people have spiders which result in duplicate KVS names. If they do, we can hash the original spider name and append a piece of that hash to the KVS name, but for now that felt like an overkill. Let me know what you think.

honzajavorek replied:
Just a note, I've read the Scrapy docs and the spider name doesn't seem to be limited in any way to certain sets of characters. It's just an arbitrary unicode string. If we later find issues in my implementation, I'd opt for bringing `python-slugify` in, because I wouldn't like to re-implement it here, that library is battle-tested.

vdusek replied:
This is good enough, thanks.

vdusek commented on src/apify/scrapy/extensions/_httpcache.py:
```diff
@@ -0,0 +1,190 @@
+            config = Configuration.get_global_configuration()
+            if config.is_at_home:
+                storage_client = ApifyStorageClient.from_config(config)
+                return await KeyValueStore.open(name=kvs_name, storage_client=storage_client)
```
This means indefinite data retention of this storage, which is undesirable in the case of a cache.

https://docs.apify.com/platform/storage/usage#data-retention

honzajavorek replied:
Correct, that's why there's a "garbage collector" in `close_spider()`, which gradually removes expired items.

honzajavorek replied:
That page mentions another interesting limit though:

> Storages' names can be up to 63 characters long.

vdusek replied:
Okay, thanks.

> That page mentions another interesting limit though:
> Storages' names can be up to 63 characters long.

Maybe could you make sure in the `get_kvs_name` that the name isn't longer? I guess it's okay just slice off the rest.

honzajavorek replied:
Done in [22af7ac](https://github.com/apify/apify-sdk-python/pull/403/commits/22af7acf189712d4e2ff7fb80a52c9af22f84d51)

vdusek reviewed (approved):
LGTM, thanks

honzajavorek commented:
![cat-highfive](https://github.com/user-attachments/assets/5cf463d2-1cb7-44ad-afb7-ecfb9decedf1)
